[spring-ecommerce-order - Step-1] Submit deniz-oezdemir's mission

deniz-oezdemir opened the pull request:
Hey there 👋

I hope you’re doing well.

I have two questions I’d like your input on:
- This is our first Spring-based project that uses JPA. Are we following best practices? How could we make the code easier to maintain as we develop further?
- With guidance from the instructors, we implemented the `subtract` method for the `Option` class, but we are not using it yet. Our understanding is that in the future, when we have product stock with different options, we will need to subtract from the relevant option’s stock when a customer adds it to the cart. Is that correct? 🤔

Thanks for taking the time to review this. 🙏

Have a great weekend 😊

All the best,
Deniz

woowahan-pjs committed:
Initial commit

woowahan-pjs committed:
feat: set up the project

SebasNadu committed:
docs: update README with project features and roadmap

SebasNadu committed:
feat: add Product model with update and toEntity methods

SebasNadu committed:
feat: add ProductController with CRUD endpoints for products

SebasNadu committed:
feat: add NotFoundException for handling missing resources

SebasNadu committed:
feat: implement global exception handler and enhance ProductController initialization

SebasNadu committed:
feat: implement global exception handler and enhance ProductController initialization

SebasNadu committed:
refactor: remove unused @ControllerAdvice from ProductController

SebasNadu committed:
chore: update Gradle wrapper jar file

SebasNadu committed:
refactor: update annotations in ProductController and improve code formatting

SebasNadu committed:
refactor: update annotations in ProductController and improve code formatting

SebasNadu committed:
docs: update README with additional roadmap steps and milestones

SebasNadu committed:
refactor: remove unused @RestController import from ProductController

SebasNadu committed:
refactor: update ProductController endpoint mappings for consistency

SebasNadu committed:
test: add initial structure for ProductControllerTest with placeholder methods

SebasNadu committed:
docs: update README with new tasks for GlobalControllerAdvise and ProductService

SebasNadu committed:
feat: implement ProductService with CRUD operations and preload sample data

SebasNadu committed:
feat: refactor ProductController to use ProductService and add ProductViewController for UI

SebasNadu committed:
test: add ProductControllerTest with comprehensive CRUD operation tests

SebasNadu committed:
feat: add ProductViewController and product-list.html for displaying products

SebasNadu committed:
feat: enable Spring DevTools for improved development experience with hot updates

SebasNadu committed:
feat: update product-list.html to display product names and prices in a list

SebasNadu committed:
refactor: clean up code formatting and improve readability in Product and ProductController classes

SebasNadu committed:
feat: add deleteAllProducts endpoint to ProductController and supporting service method

SebasNadu committed:
feat: enhance product list UI with Bootstrap styling and table layout

SebasNadu committed:
feat: implement single product deletion functionality using JS

SebasNadu committed:
feat: implement single product deletion functionality using JS

Commit:
feat: implement delete, add, edit product using JS

Commit:
feat: implement delete, add, edit product using JS

Commit:
chore: removed unnecessary comments

Commit:
add(application.properties): h2 setup

Commit:
add(application.properties): h2 setup

Commit:
fix(*.sql): move sql files to right location

Commit:
fix(*.sql): move sql files to right location

Commit:
chore: replace imageUrl with real pictures' URLs

Commit:
chore: add TODO

Commit:
chore(product-list.html): display image instead of url

Commit:
chore(data.sql): replace image urls to the valid ones

Commit:
add: add configuration for h2 database

Commit:
add: replace product service with product repository to use h2 database

Commit:
add: update methods in the new class product repository to use h2 database

Commit:
fix: throw appropriate errors and incorporate modified database to test

Commit:
docs(README.md): update feature list

SebasNadu committed:
chore: update application properties with detailed configs and remove unused static file placeholder

SebasNadu committed:
refactor: standardize table name to "PRODUCT" across schema, data, and repository

SebasNadu committed:
refactor: create a ProductRepository interface and implement it in a jdbc and collection repository. Update references in controllers

SebasNadu committed:
refactor: create a ProductRepository interface and implement it in a jdbc and collection repository. Update references in controllers

SebasNadu committed:
refactor: create a ProductRepository interface and implement it in a jdbc and collection repository. Update references in controllers

SebasNadu committed:
feat: add Font Awesome icons for delete and edit buttons, and implement product update modal

SebasNadu committed:
refactor: update formatting and improve readability in ProductRepository and its implementation

SebasNadu committed:
refactor: update formatting and improve readability in ProductRepository and its implementation

SebasNadu committed:
docs: update README to reflect changes in product repository and introduce ProductDTO structure

SebasNadu committed:
feat: add Spring Boot starter validation dependency for improved input validation

SebasNadu committed:
feat: introduce ProductDTO and ProductPatchDTO for improved data handling

SebasNadu committed:
feat: implement ProductService interface and its collection and JDBC implementations for product management

SebasNadu committed:
feat: implement ProductService interface and its collection and JDBC implementations for product management

SebasNadu committed:
feat: add ProductMapper for converting between Product and ProductDTO

SebasNadu committed:
feat: implement ProductRepository for CRUD operations on Product entities

SebasNadu committed:
feat: add custom exceptions and global exception handling for improved error management

SebasNadu committed:
feat: add custom exceptions and global exception handling for improved error management

SebasNadu committed:
feat: refactor ProductController and ProductViewController to use ProductService for improved product management

SebasNadu committed:
feat: update application properties for livereload and add gradle daemon JVM properties

SebasNadu committed:
feat: enhance product management UI with modal for adding new products and update ProductViewController for product creation

SebasNadu committed:
feat: implement global error handling with custom error pages and exception advice

SebasNadu committed:
feat: implement global error handling with custom error pages and exception advice

SebasNadu committed:
feat: refactor product management to use modals for adding and editing products and update product deletion logic

SebasNadu committed:
feat: implement pagination for product listing and update ProductViewController to support paginated results

SebasNadu committed:
feat: update product pagination logic and enhance layout in product-list.html

SebasNadu committed:
feat: update README with completed tasks for ProductDTO, ProductService, and pagination implementation

SebasNadu committed:
refactor: reformat method signatures and align parameter declarations with standard code style

SebasNadu committed:
docs(readme): complete Product Name validation in README checklist

SebasNadu committed:
feat: implement product validation, enhance error handling, and complete repository refactor

- Added validation to `ProductDTO` fields (name, price, imageUrl) using Jakarta annotations.
- Enhanced error handling for API and View with detailed validation, JDBC, and exception responses.
- Refactored `ProductRepository` to an interface, added `ProductRepositoryImpl` for JDBC implementation.
- Updated `ProductService` to include product name uniqueness validation.
- Augmented tests for validation errors (name constraints, duplicate names, prices, etc.).

SebasNadu committed:
docs(readme): mark product validation checklist items as completed

SebasNadu committed:
[spring-ecommerce-product Step 1] Submit SebasNadu's mission (#13)

* docs: update README with project features and roadmap

* feat: add Product model with update and toEntity methods

* feat: add ProductController with CRUD endpoints for products

* feat: add NotFoundException for handling missing resources

* feat: implement global exception handler and enhance ProductController initialization

* refactor: remove unused @ControllerAdvice from ProductController

* chore: update Gradle wrapper jar file

* refactor: update annotations in ProductController and improve code formatting

* docs: update README with additional roadmap steps and milestones

* refactor: remove unused @RestController import from ProductController

* refactor: update ProductController endpoint mappings for consistency

* test: add initial structure for ProductControllerTest with placeholder methods

* docs: update README with new tasks for GlobalControllerAdvise and ProductService

* feat: implement ProductService with CRUD operations and preload sample data

* feat: refactor ProductController to use ProductService and add ProductViewController for UI

* test: add ProductControllerTest with comprehensive CRUD operation tests

* feat: add ProductViewController and product-list.html for displaying products

* feat: enable Spring DevTools for improved development experience with hot updates

* feat: update product-list.html to display product names and prices in a list

* refactor: clean up code formatting and improve readability in Product and ProductController classes

* feat: add deleteAllProducts endpoint to ProductController and supporting service method

* feat: enhance product list UI with Bootstrap styling and table layout

* feat: implement single product deletion functionality using JS

* feat: implement delete, add, edit product using JS

* chore: removed unnecessary comments

* add(application.properties): h2 setup

* fix(*.sql): move sql files to right location

* chore: replace imageUrl with real pictures' URLs

* chore: add TODO

* chore(product-list.html): display image instead of url

* chore(data.sql): replace image urls to the valid ones

* add: add configuration for h2 database

* add: replace product service with product repository to use h2 database

* add: update methods in the new class product repository to use h2 database

* fix: throw appropriate errors and incorporate modified database to test

* docs(README.md): update feature list

* chore: update application properties with detailed configs and remove unused static file placeholder

* refactor: standardize table name to "PRODUCT" across schema, data, and repository

* refactor: create a ProductRepository interface and implement it in a jdbc and collection repository. Update references in controllers

* refactor: create a ProductRepository interface and implement it in a jdbc and collection repository. Update references in controllers

* feat: add Font Awesome icons for delete and edit buttons, and implement product update modal

* refactor: update formatting and improve readability in ProductRepository and its implementation

* docs: update README to reflect changes in product repository and introduce ProductDTO structure

* feat: add Spring Boot starter validation dependency for improved input validation

* feat: introduce ProductDTO and ProductPatchDTO for improved data handling

* feat: implement ProductService interface and its collection and JDBC implementations for product management

* feat: add ProductMapper for converting between Product and ProductDTO

* feat: implement ProductRepository for CRUD operations on Product entities

* feat: add custom exceptions and global exception handling for improved error management

* feat: refactor ProductController and ProductViewController to use ProductService for improved product management

* feat: update application properties for livereload and add gradle daemon JVM properties

* feat: enhance product management UI with modal for adding new products and update ProductViewController for product creation

* feat: implement global error handling with custom error pages and exception advice

* feat: refactor product management to use modals for adding and editing products and update product deletion logic

* feat: implement pagination for product listing and update ProductViewController to support paginated results

* feat: update product pagination logic and enhance layout in product-list.html

* feat: update README with completed tasks for ProductDTO, ProductService, and pagination implementation

---------

Co-authored-by: Jin Park <jieun.park.e@deliveryhero.com>

SebasNadu committed:
feat: add form validation and error handling for product creation and update modals

- Integrated field-level validation in `ProductDTO` and `ProductPatchDTO` using Jakarta annotations.
- Enhanced Thymeleaf templates to display validation errors in modals.
- Updated JavaScript to automatically display the modal when validation errors occur.
- Improved backend validation error handling and feedback for `ProductController` and `ProductViewController`.

SebasNadu committed:
docs(readme): mark ProductDTO validation and feedback tasks as completed

SebasNadu committed:
refactor: improve README grammar, adjust validation annotations in ProductPatchDTO, and fix minor formatting issues

- Corrected grammatical errors and improved clarity in `README.md`.
- Removed redundant `@NotBlank` annotations in `ProductPatchDTO`, retaining necessary constraints.
- Fixed minor code formatting inconsistencies in `ProductViewController` and `build.gradle.kts`.

SebasNadu committed:
docs(readme): document initial checklist for Member entity and JWT integration tasks

SebasNadu committed:
feat: add Member entity, DTO, and mapper

- Created `Member` entity with `Role` enum for role differentiation.
- Added `MemberDTO` with field-level Jakarta validation.
- Implemented `MemberMapper` for entity-DTO conversion.

SebasNadu committed:
docs(readme): update checklist to mark Member entity, DTO, and mapper tasks as completed

SebasNadu committed:
chore(schema): update schema to add Member table and adjust Product constraints

SebasNadu committed:
feat: add TokenRequestDTO and TokenResponseDTO for authentication

- Introduced `TokenRequestDTO` with field-level validation for email and password.
- Added `TokenResponseDTO` to structure access token responses.
- Updated `README.md` checklist to reflect completed DTO implementations.

SebasNadu committed:
feat: implement JWT token infrastructure with token provider and extractor

- Added `JwtTokenProvider` to create, validate, and parse JWT tokens.
- Introduced `AuthorizationExtractor` to extract bearer tokens from request headers.
- Updated `README.md` checklist to mark JWT-related tasks as completed.

SebasNadu committed:
feat: implement MemberRepository and MemberService with JDBC integration

- Added `MemberRepository` interface and implemented it using `JdbcTemplate`.
- Introduced `MemberService` interface and partially implemented `MemberServiceImpl`.
- Enabled member CRUD operations and email uniqueness validation.

SebasNadu committed:
docs(readme): mark MemberService and repository tasks as completed

SebasNadu committed:
feat: implement AuthService and MemberController with JWT integration

- Added `AuthService` to handle authentication, token generation, and validation.
- Introduced `MemberController` with register, login, and findMe endpoints.
- Completed corresponding tests in `MemberControllerTest`.
- Updated `README.md` checklist to mark related tasks as completed.

SebasNadu committed:
feat: add JWT security properties and JJWT dependencies

- Added JWT secret key and expiration length in `application.properties`.
- Included `JJWT` library dependencies in `build.gradle.kts` for token handling.

SebasNadu committed:
feat: add custom exceptions for authorization errors and update global handlers

- Added `AuthorizationException` and `ForbiddenException` for handling authorization errors.
- Updated `ViewErrorControllerAdvice` and `ApiErrorControllerAdvice` to handle new exceptions.
- Updated `README.md` checklist to mark related tasks as completed.

SebasNadu committed:
refactor: remove unused JavaScript utilities and cleanup ProductPatchDTO

- Deleted `main.js` as its functionality was no longer required.
- Simplified imports in `ProductPatchDTO` by removing unnecessary `@NotBlank` constraint.

SebasNadu committed:
docs(readme): update checklist with tasks for WebMvcConfig, admin user creation, and tests

SebasNadu committed:
feat: add WebMvcConfiguration with login check interceptor

- Implemented `WebMvcConfiguration` to register `CheckLoginInterceptor`.
- Created `CheckLoginInterceptor` for JWT validation and role checks.
- Updated `ViewErrorControllerAdvice` to redirect unauthorized/forbidden errors to the login page.
- Marked the associated tasks as completed in `README.md`.

SebasNadu committed:
feat(data.sql): add admin user insertion script

SebasNadu committed:
test: add token-based authentication to controller tests

- Integrated login and token retrieval in `BeforeEach` setup.
- Updated `ProductControllerTest` and `MemberControllerTest` to include token in API requests.

SebasNadu committed:
docs(readme): mark admin user creation and test updates as complete

SebasNadu committed:
feat(member): update mappers to include role in DTO and entity conversion

SebasNadu committed:
feat(member): add `enrichedWithRole` function and refactor formatting in repository/service files

- Introduced `enrichedWithRole` method in `MemberService` and its implementation to include role enrichment in `MemberDTO`.
- Refactored formatting in `MemberRepositoryImpl` and `MemberServiceImpl` for improved readability.

SebasNadu committed:
refactor(auth): enhance role handling in JWT and tighten interceptor validation

- Updated JWT creation to include role as a claim and adjusted payload parsing accordingly.
- Refactored `AuthService` to replace `TokenRequestDTO` with `MemberDTO` for better consistency.
- Limited `CheckLoginInterceptor` to validate only `POST` operations and added admin role checks.
- Modified path patterns in `WebMvcConfiguration` to apply login checks only on `/api/products/**`.

SebasNadu committed:
refactor(advice): adjust formatting and remove unused imports in error handlers

- Improved readability of `ExceptionHandler` methods in `ViewErrorControllerAdvice` and `ApiErrorControllerAdvice` by updating formatting.
- Removed unused `Instant` import from `ViewErrorControllerAdvice`.

SebasNadu committed:
test(controller): improve formatting and ensure token inclusion in API tests

- Refactored code for better readability in `ProductControllerTest` and `MemberControllerTest`.
- Ensured token is included in all API requests for proper authentication.

SebasNadu committed:
docs(readme): clarify interceptor implementation in WebMvcConfig checklist

SebasNadu committed:
docs(readme): update checklist for Step 2.3 with CartItem and admin interceptor details

SebasNadu committed:
feat(cart): add CartItem entity and request/response DTOs with database schema

SebasNadu committed:
feat(cart): implement CartItemController and related services for cart operations

SebasNadu committed:
feat(config): add role-based interceptor, argument resolver, and enhancements for authorization

- Introduced `CheckRoleInterceptor` for role-based request validation.
- Added `LoginMemberArgumentResolver` to resolve `@LoginMember` in controllers.
- Enhanced `CheckLoginInterceptor` to handle ignored methods/annotations and include role extraction.
- Updated `WebMvcConfiguration` to integrate new interceptors and argument resolvers.
- Created `@LoginMember` annotation for method parameter injection.

SebasNadu committed:
feat(interceptor): add CheckRoleInterceptor and CheckAdminOnly annotation for role-based access control

SebasNadu committed:
refactor(e2e): rename and restructure E2E test classes for products and members

- Moved `ProductControllerTest` and `MemberControllerTest` to the `ecommerce.etoe` package.
- Renamed classes to `ProductE2ETest` and `MemberE2ETest`, respectively, for better alignment with their purpose.

refactor(member): rename DTO mapper methods for clarity

- Renamed `toDto` to `toResponseDto` across `MemberMapper` and dependent service implementations.
- Improved method naming to convey the intent of converting to a response format.

refactor(product): use snake_case for database column names

- Updated `ProductRepositoryImpl` to use `image_url` instead of `imageUrl` in queries and mappings.

feat(controller): add role-based access control annotations

- Introduced `@CheckAdminOnly` and `@IgnoreCheckLogin` annotations in `ProductController` for method-level access control.

SebasNadu committed:
feat(admin): add AdminController and service for managing admin operations

SebasNadu committed:
feat(admin): complete admin operations with service and role enhancements

- Implemented `AdminController` with endpoints for top products and active members.
- Added `AdminService` with methods to fetch top products and active members data.
- Updated `CheckRoleInterceptor` to handle `@IgnoreCheckLogin` annotation and throw `ForbiddenException` for unauthorized access.
- Updated README.md checklist reflecting the completion of admin-related tasks.
- Added integration tests for `AdminService` to verify functionality.

SebasNadu committed:
feat(config): refine interceptors and seed data for enhanced role-based control

- Updated `WebMvcConfiguration` to limit paths in `CheckLoginInterceptor` and `CheckRoleInterceptor`.
- Expanded seed data in `data.sql` with additional products, members, and cart items.
- Added `ActiveMemberDTO` for new member-related functionality.

SebasNadu committed:
feat(e2e): add AdminE2ETest and rename E2E test packages for consistency

- Added `AdminE2ETest` to cover admin endpoints for top-products and active-members.
- Renamed `ecommerce.etoe` package to `ecommerce.endtoend` for improved clarity.
- Updated assertions in `ProductE2ETest` for validating increased product count.
- Updated README checklist with new integration test task.

SebasNadu committed:
feat(test): add integration tests for CartItem, Member, and Product services

- Implemented `CartItemServiceTest` to validate cart item operations.
- Added `MemberServiceTest` to cover member-related functionalities.
- Created `ProductServiceJDBCTest` for testing product operations via JDBC.
- Updated README checklist to mark integration tests as complete.

SebasNadu committed:
feat(cart, member): add `deleteAll` methods and enhanced repository queries

- Added `deleteAll` methods in services and repositories for CartItem and Member.
- Enhanced `CartItemRepository` with queries for top products and active members.
- Updated `existsByProduct` logic to validate against both product and member.

SebasNadu committed:
feat(auth, product): add token-based authentication for product updates and deletion

- Introduced `getCookie` util function to extract token from cookies.
- Added `Authorization` header with Bearer token in `updateProduct` and `deleteProductById` requests.
- Updated product views with `@IgnoreCheckLogin` annotation for selective login bypass.

SebasNadu committed:
refactor(test): remove unused import `CartItemRepository` from `CartItemServiceTest`

SebasNadu committed:
Merge branch 'sebasnadu' into step2

SebasNadu committed:
test: remove `ProductControllerTest` and clean up unused imports in `ApiErrorControllerAdvice`

SebasNadu committed:
refactor: simplify product list loading in ProductViewController

SebasNadu committed:
feat: add validation messages for DTOs and refactor validation annotations

SebasNadu committed:
feat: enhance validation messages in DTOs and refactor authorization token extraction

SebasNadu committed:
refactor: optimize SQL query for fetching distinct members with recent cart activity and add indexes for performance

SebasNadu committed:
refactor: remove unnecessary DirtiesContext annotations and update test assertions for consistency

SebasNadu committed:
refactor: remove unused imports in ProductE2ETest

SebasNadu committed:
feat: add name field to Member and ActiveMemberDTO, update related services and tests

SebasNadu committed:
test: add unit tests for AuthorizationExtractor and JwtTokenProvider

SebasNadu committed:
refactor: clean up imports in AuthorizationExtractorTest and JwtTokenProviderTest

SebasNadu committed:
chore(config): remove unused H2 database configuration properties from `application.properties`

SebasNadu committed:
refactor: clean up imports in AuthorizationExtractorTest and JwtTokenProviderTest

SebasNadu committed:
refactor: update indexes in schema.sql for cart_item and product tables

SebasNadu committed:
Merge remote-tracking branch 'origin/step2' into step2

SebasNadu committed:
docs: update README to replace product-related documentation with order-related tasks

- Updated project title to reflect focus on order management.
- Removed product-related steps in favor of JPA integration for order handling.
- Added tasks for replacing `JdbcTemplate` with JPA and logging JPA operations.

SebasNadu committed:
feat(config): update JPA and logging configurations, add `spring-boot-starter-data-jpa`

- Modified `application.properties` to adjust JPA settings, enable SQL formatting, and trace JDBC bindings.
- Added `spring-boot-starter-data-jpa` to dependencies in `build.gradle.kts`.
- Enabled `kotlin("plugin.jpa")` for JPA-related enhancements.

SebasNadu committed:
refactor(product): migrate from `JdbcTemplate` to JPA for product-related operations

- Replaced `ProductRepositoryImpl` and `ProductServiceJDBC` with JPA-based implementations.
- Updated `Product` entity to include JPA annotations.
- Removed outdated pagination logic and marked it as TODO.
- Renamed and updated integration test class to `ProductServiceTest`.
- Cleaned up unused test methods and adjusted exception handling.

SebasNadu committed:
refactor(member): migrate from `JdbcTemplate` to JPA for member-related operations

- Replaced `MemberRepositoryImpl` with JPA-based implementation.
- Updated `Member` entity with JPA annotations.
- Removed unused custom repository methods and simplified transaction management.
- Adjusted exception handling in `MemberServiceImpl`.
- Renamed `toResponseDto` to `toDto`.

SebasNadu committed:
refactor(cart): migrate `CartItem` operations from `JdbcTemplate` to JPA

- Replaced `CartItemRepositoryImpl` with JPA-based implementation.
- Updated `CartItem` entity with JPA annotations, added relationships with `Member` and `Product`.
- Modified service methods to accept `MemberDTO` instead of `memberId`.
- Simplified repository queries and removed redundant custom methods.
- Adjusted related tests to align with JPA changes, added `@Transactional` where needed.

SebasNadu committed:
refactor(admin): update E2E and integration tests with JPA improvements and DTO changes

- Added `@Transactional` to `AdminE2ETest` and `AdminServiceTest`.
- Adjusted `TopProductDTO` to interface with updated fields and structure.
- Replaced direct `memberId` usage with `member.toDto()` in tests.
- Marked `AdminServiceImpl` as `@Primary`.
- Modified JSON path in E2E test to adapt to updated query response structure.

SebasNadu committed:
chore(schema): reorder `DROP TABLE` statements in `schema.sql` to fix dependency issue

SebasNadu committed:
chore(schema): update `schema.sql` to use `InnoDB` and enhance H2 config for MySQL compatibility

SebasNadu committed:
chore(schema): add `WISH_ITEM` table and update `schema.sql` FK constraints

- Added `WISH_ITEM` table with fields, constraints, and indexes.
- Renamed FK constraints for `CART_ITEM` table to improve clarity.
- Reformatted SQL statements for consistency.

SebasNadu committed:
feat(wishlist): add `WishItem` feature with CRUD operations and E2E tests

- Introduced `WishItem` entity, repository, service, and controller for managing wishlist items.
- Added DTOs and mapper for `WishItem` conversions.
- Implemented CRUD operations: add, retrieve, and delete wish items.
- Created end-to-end tests to ensure functionality and API correctness.

SebasNadu committed:
docs: mark JPA migration tasks as complete in README

- Updated checklist items for replacing `JdbcTemplate` with JPA and logging JPA operations to completed.

SebasNadu committed:
docs: update README with pagination and wishlist progress

- Added new checklist items for pagination support in products and wishlist.
- Marked items related to these tasks in the README as completed.

SebasNadu committed:
feat(product): implement pagination for product listing with DTO and JPA support

- Introduced `PageResponseDTO` for handling paginated responses.
- Updated `ProductService` to return paginated results using `Pageable`.
- Refactored `ProductController` and `ProductViewController` to support and use pagination.
- Modified E2E and integration tests to validate paginated product responses.
- Removed outdated custom pagination logic and related methods.

SebasNadu committed:
refactor(product): remove obsolete `ProductServiceCollection` implementation

- Deleted the old in-memory `ProductServiceCollection` class.
- Cleanup aligns with the migration to JPA-based product services.

SebasNadu committed:
feat(wishlist): add pagination support for wishlist retrieval

- Updated `WishItemService` and `WishItemRepository` to support paginated responses.
- Modified `WishItemController` to handle pagination using `Pageable`.
- Adjusted DTO mappings and added `PageResponseDTO` type in E2E tests.
- Refactored existing methods to improve clarity and consistency.

SebasNadu committed:
feat(member): add `wishItems` association to `Member` entity

- Introduced `wishItems` field in `Member` entity with proper JPA mappings.
- Updated `MemberDTO` to align with the new association (commented for now).

SebasNadu committed:
refactor(controller, test): fix variable naming in `WishItemController` and improve `CartItemE2ETest`

- Corrected variable naming in `WishItemController` for consistency with Kotlin naming conventions.
- Simplified `CartItemE2ETest` by removing unused `addCartItem` method and introducing `@TestInstance` for optimized lifecycle management.
- Applied minor formatting fix in `schema.sql`.

SebasNadu committed:
docs: update README with Step 1.3 tasks for option feature development

- Added new checklist items in Step 1.3 for implementing the option feature, including validations, API enhancements, and admin endpoint.

SebasNadu committed:
refactor(wishlist): replace `ProductDTO` with `ProductResponseDTO` for consistency

- Updated `WishItemResponseDTO` to use `ProductResponseDTO` type.
- Adjusted `WishItemE2ETest` assertions and mappings to reflect the new type.
- Removed unused imports in `WishItemController`.

SebasNadu committed:
feat(database): migrate SQL scripts to Kotlin-based database initialization

- Removed `.sql` schema and data files in favor of Kotlin configuration for database setup.
- Added `DatabaseConfig` to seed initial data, including products, members, cart items, and options.
- Simplified entity relationships and initial data setup using JPA repositories.

SebasNadu committed:
feat(option): add `Option` entity and corresponding `OptionDTO`

- Introduced `Option` entity with fields for name, quantity, and product association, including validations for name and quantity constraints.
- Added `OptionDTO` to handle data transfer with appropriate validation annotations.

SebasNadu committed:
refactor(product): replace `ProductDTO` with `ProductRequestDTO` and `ProductResponseDTO`

- Introduced `ProductRequestDTO` and `ProductResponseDTO` for better request/response separation.
- Updated controllers, services, mappers, and tests to use the new DTOs.
- Enhanced `Product` entity to support `options` with validation and updated associations.

SebasNadu committed:
feat(option): add `OptionMapper` and `OptionRepository` classes

- Introduced `OptionMapper` for mapping between `Option` and `OptionDTO` objects.
- Added `OptionRepository` extending `JpaRepository` for `Option` entity persistence.

SebasNadu committed:
refactor(cart): replace `ProductDTO` with `ProductResponseDTO` in `CartItemResponseDTO` and tests

- Updated `CartItemResponseDTO` to use `ProductResponseDTO` for consistency with product response handling.
- Adjusted `CartItemE2ETest` to reflect the new type and updated relevant assertions and test mappings.

SebasNadu committed:
refactor(product): replace `ProductDTO` with `ProductRequestDTO` and `ProductResponseDTO` in `ProductService`

- Updated `ProductService` interface and implementation to use `ProductRequestDTO` for input and `ProductResponseDTO` for output.
- Enhanced `save` and `updateById` methods to handle `options` within products using the new DTOs.
- Added `OptionRepository` dependency to facilitate `options` persistence.

SebasNadu committed:
feat(exception): add custom exceptions and handlers for validation and stock errors

- Introduced new exceptions: `NoSuchElementException`, `InvalidOptionNameException`, `InvalidOptionQuantityException`, and `InsufficientStockException`.
- Added handlers in `ApiErrorControllerAdvice` for the newly created exceptions.
- Extended `ValidationMessages` with new constants for option validation messages.

SebasNadu committed:
docs: update README to reflect completed Step 1.3 tasks for option feature

- Marked tasks related to `Option` entity creation, validations, and admin functionality as complete.
- Added a subsection for future work regarding stock management for product options.

SebasNadu committed:
feat(admin): add API endpoint and service method to create product options

- Added `createOption` method in `AdminService` and its implementation.
- Introduced a new `/newOption` POST endpoint in `AdminController` for creating product options.
- Updated `Product` entity to use `InvalidOptionNameException` for duplicate option validation.

SebasNadu committed:
refactor(product): remove redundant mapping of options in `ProductServiceImpl`

SebasNadu committed:
feat(admin): add E2E tests for product option creation and validation

- Added `AdminE2ETest` methods to test the creation of product options and validate constraints:
  - Rejects options with invalid quantities (<1 or ≥100,000,000).
  - Ensures unique option names per product.
  - Validates option name length limits.
- Updated setup to fetch product details before running tests.
- Refactored tests to use `ProductRequestDTO` for consistent input handling.

SebasNadu committed:
refactor: standardize formatting across controllers, services, and data classes

- Added missing trailing commas to parameter lists and method signatures.
- Adjusted indentation for lambda expressions and map constructions.
- Removed commented-out code and unused imports.
- Ensured consistent annotation and field placements in DTOs and entities.

SebasNadu committed:
refactor: optimize the service methods with @Transactional method

SebasNadu committed:
feat(product): enhance product patching with options and improve DTO handling

SebasNadu committed:
refactor: remove unused imports and clean up repository and DTO files

SebasNadu committed:
refactor: change data classes to regular classes and update quantity handling in cart items

SebasNadu committed:
Merge branch 'step1' into step1-fixed

SebasNadu committed:
chore: remove TODO comment from application.properties

Wordbe commented on src/main/kotlin/ecommerce/advice/ApiErrorControllerAdvice.kt:
```diff
@@ -38,6 +41,20 @@ class ApiErrorControllerAdvice {
+    fun handleNotSuchElementException(e: NoSuchElementException): ResponseEntity<Map<String, Any>> {
+        val errorMessage = e.message ?: "Not such element"
+        log.warn("NoSuchElementException occurred: $errorMessage", e)
+        val body =
+            mapOf(
+                "status" to HttpStatus.NOT_FOUND.value(),
+                "error" to "Operation failed",
+                "message" to errorMessage,
+                "timestamp" to Instant.now(),
+            )
+        return ResponseEntity.status(HttpStatus.NOT_FOUND).body(body)
```
It would be great to have a dedicated error object for consistent error responses! 🙂

deniz-oezdemir replied:
Thanks for the great idea! I created a dedicated `ApiError` class to make error responses consistent across the app. Now all exceptions return errors in this format, and validation errors include a clear details field. This should make the API easier to use and maintain. 😊

Wordbe replied:
Sounds good!

Wordbe commented on src/main/kotlin/ecommerce/advice/ApiErrorControllerAdvice.kt:
```diff
@@ -94,6 +111,48 @@ class ApiErrorControllerAdvice {
+                "message" to errorMessage,
+                "timestamp" to Instant.now(),
+            )
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST).body(body)
```
It also seems like it could be reusable.

deniz-oezdemir replied:
Thanks, noted! See my reply above. 🙂

Wordbe commented on src/main/kotlin/ecommerce/advice/ApiErrorControllerAdvice.kt:
```diff
@@ -94,6 +111,48 @@ class ApiErrorControllerAdvice {
+                "error" to "Invalid option quantity",
+                "message" to errorMessage,
+                "timestamp" to Instant.now(),
+            )
```
Same here.

deniz-oezdemir replied:
Thanks, noted! See my reply above. 🙂

Wordbe commented on src/main/kotlin/ecommerce/advice/ApiErrorControllerAdvice.kt:
```diff
@@ -94,6 +111,48 @@ class ApiErrorControllerAdvice {
+                "error" to "Insufficient stock",
+                "message" to errorMessage,
+                "timestamp" to Instant.now(),
+            )
```
Same here.

deniz-oezdemir replied:
Thanks, noted! See my reply above. 🙂

Wordbe commented on src/main/kotlin/ecommerce/controller/AdminController.kt:
```diff
@@ -17,4 +21,12 @@ class AdminController(private val adminService: AdminService) {
     @GetMapping("/active-members")
     fun getActiveMembers(): List<ActiveMemberDTO> = adminService.findMembersWithRecentCartActivity()
+
+    @PostMapping("/newOption")
```
Try to standardize the URL naming convention for REST APIs across the project. 🙂

deniz-oezdemir replied:
Thanks for the feedback! I standardized the URIs as suggested and fixed the tests accordingly. Let me know if anything else needs adjustment. 🙂

Wordbe replied:
Great job.

Wordbe commented on src/main/kotlin/ecommerce/model/OptionDTO.kt:
```diff
@@ -0,0 +1,22 @@
+    @field:Size(min = 1, max = 99_999_999, message = OPTION_QUANTITY_SIZE)
+    val quantity: Long,
+    @field:NotBlank(message = OPTION_PRODUCT_ID_REQUIRED)
+    val productId: Long? = null,
```
Is there a specific reason you're accepting this as nullable? 🤔

deniz-oezdemir replied:
Thanks for the question! The `productId` is nullable because when creating a product, it should not be provided as the product and its options are created together in that flow.
For standalone option creation, I now explicitly validate that `productId` is present and return a clear error if it’s missing. I also removed the incorrect `@NotBlank` annotation (since it doesn’t apply to a Long) and handle the requirement in the admin endpoint logic instead.
For the scope of this project, I think a single DTO is sufficient, but I’m open to discussing separate DTOs if you think it would improve clarity. 🙂

Wordbe replied:
Oh, it’s also used when creating a product. In that case, yes — it seems fine to use a single DTO for both cases. 👍

Wordbe commented on src/main/kotlin/ecommerce/controller/ProductController.kt:
```diff
@@ -29,37 +28,37 @@ class ProductController(private val productService: ProductService) {
     @GetMapping(PRODUCT_PATH)
     fun getProducts(
         @PageableDefault(size = 10, sort = ["name"], direction = Sort.Direction.ASC)
```
Consider extracting the magic number (`"name"`) into a constant. It would be especially helpful to define the allowed range for sorting options. 🙂

deniz-oezdemir replied:
Good point, I’ve moved `name` into a `DEFAULT_SORT_FIELD` constant and also introduced an `ALLOWED_SORT_FIELDS` set (name, price). I now validate the incoming sort fields in `getProducts` and fall back to the default when an unsupported field is passed. This keeps the behavior predictable and makes it easier to expand/maintain allowed sort options. Thanks for pointing that out! 🙏

Wordbe commented on src/main/kotlin/ecommerce/controller/ProductViewController.kt:
```diff
@@ -23,28 +23,31 @@ class ProductViewController(private val productService: ProductService) {
         model.addAttribute("currentPage", products.number)
         model.addAttribute("totalPages", products.totalPages)
-        model.addAttribute("productDTO", ProductDTO(null, "", 0.0, ""))
+        model.addAttribute("productDTO", ProductResponseDTO(null, "", 0.0, ""))
```
Why is this being set as an empty value here? 🙂

deniz-oezdemir replied:
We switched it to `ProductResponseDTO` because the view uses that for form binding in Thymeleaf (`th:object="${productDTO}"`). This way, the form fields line up perfectly with what the view expects, even when we’re just showing an empty product for creation. It’s basically just a placeholder so the form has something to bind to. Good catch! 😊

Wordbe replied:
Then what about creating a static method like `ProductResponseDTO.empty()`? 🙂

deniz-oezdemir replied:
Yes, I agree this `empty()` method makes the code cleaner and more readable.

Wordbe commented on src/main/kotlin/ecommerce/entities/CartItem.kt:
```diff
@@ -28,7 +28,7 @@ data class CartItem(
     val product: Product,
     @Column(name = "quantity", nullable = false)
-    val quantity: Int,
+    var quantity: Int,
```
Instead of using the setter directly, how about using a method with a clear name? 🙂 Let’s also think about why that approach would be better! 💡

deniz-oezdemir replied:
Yes, I replaced the direct quantity assignment with a dedicated `updateQuantity()` method on `CartItem`. This improves encapsulation and makes it easier to add validation or other logic whenever the quantity changes. 😊

Wordbe replied:
Good Job!

Wordbe commented on src/main/kotlin/ecommerce/entities/Option.kt:
```diff
@@ -0,0 +1,66 @@
+        set(value) {
+            validateQuantity(value)
+            field = value
+        }
```
Let’s create a new update method instead of using the setter, and perform validation inside it. 🙂

deniz-oezdemir replied:
Thanks for the suggestion! I replaced the direct property setters with clearly named `updateName` and `updateQuantity` methods that include validation. This makes the code more explicit and easier to follow, and prevents invalid state changes outside of those methods. 😄

Wordbe commented on src/main/kotlin/ecommerce/entities/Option.kt:
```diff
@@ -0,0 +1,66 @@
+
+    fun subtract(quantity: Long) {
+        if (quantity < 1) throw InvalidOptionQuantityException("Subtract amount must be >= 1")
+        if (this.quantity < quantity) throw InsufficientStockException("Not enough stock")
```
Good validation 👍

deniz-oezdemir replied:
Thank you!

Wordbe commented on src/main/kotlin/ecommerce/entities/Option.kt:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    private fun validateQuantity(quantity: Long) {
+        if (quantity < 1 || quantity >= 100_000_000) throw InvalidOptionQuantityException("Quantity must be between 1 and 99,999,999")
```
Please convert the magic number into a constant!

deniz-oezdemir replied:
Thanks for pointing that out! I moved the magic numbers into named constants inside a companion object (`MAX_NAME_LENGTH`, `MIN_QUANTITY`, and `MAX_QUANTITY`) to make the code clearer and easier to maintain. 🙂

Wordbe commented on src/main/kotlin/ecommerce/entities/Product.kt:
```diff
@@ -16,11 +22,59 @@ class Product(
     @OneToMany(mappedBy = "product", cascade = [CascadeType.ALL], orphanRemoval = true)
     val cartItems: Set<CartItem> = emptySet(),
-)
+    @OneToMany(cascade = [CascadeType.ALL], orphanRemoval = true)
```
Why did you choose `Cascade.ALL`? 🙂

deniz-oezdemir replied:
I chose `Cascade.ALL` because the `options` and `cartItems` are tightly coupled entities to the product. When you save, update, or delete a product, you usually want the related options and cart items to change along with it. This helps keep the data consistent and avoids leaving behind orphaned records. Using cascade here lets us manage those child entities automatically, which simplifies the code and reduces the chance of errors. Does that make sense? 🤔

Wordbe replied:
Thank you for the explanation! 😊

Wordbe commented on src/main/kotlin/ecommerce/entities/Product.kt:
```diff
@@ -16,11 +22,59 @@ class Product(
+            _options.addAll(value)
+        }
+
+    fun addOption(option: Option) {
+        if (_options.any { it.name == option.name }) {
+            throw InvalidOptionNameException("Option with name '${option.name}' already exists")
+        }
+        _options.add(option)
+    }
```
The bidirectional mapping is set up well, but I think there’s always a chance for mistakes. So it might be a good idea to include `option.product = this` when calling `addOption`. What do you think? 🙂

deniz-oezdemir replied:
Good point, I’ve updated `addOption` to set `option.product = this` before adding it to `_options`. This ensures both sides of the bidirectional mapping stay in sync. 🙏

Wordbe commented on src/main/kotlin/ecommerce/entities/Product.kt:
```diff
@@ -16,11 +22,59 @@ class Product(
+        _options.add(option)
+    }
+
+    fun copyFrom(
+        other: ProductRequestDTO,
+        optionDTOs: Set<OptionDTO> = emptySet(),
+    ) {
+        this.name = other.name
+        this.price = other.price
+        this.imageUrl = other.imageUrl
+        this.options = mapOptionDTOs(optionDTOs)
+    }
+
+    fun copyFrom(
+        other: ProductPatchDTO,
+        optionDTOs: Set<OptionDTO> = emptySet(),
+    ) {
+        other.name?.let { if (it.isNotBlank()) this.name = it }
+        other.price?.let { this.price = it }
+        other.imageUrl?.let { if (it.isNotBlank()) this.imageUrl = it }
+        if (optionDTOs.isNotEmpty()) {
+            this.options = mapOptionDTOs(optionDTOs)
+        }
+    }
+
+    private fun mapOptionDTOs(optionDTOs: Set<OptionDTO>): List<Option> {
+        return optionDTOs.map { dto ->
+            _options.find { it.id == dto.id }?.apply {
+                name = dto.name
+                quantity = dto.quantity
+            } ?: dto.toEntity(this)
+        }
+    }
```
It doesn’t seem ideal for the domain model to depend on a DTO. When the flow is one-directional in the layer structure, it becomes easier to manage dependencies. It would be better if this conversion happens externally. 🙂

deniz-oezdemir replied:
Good point! I removed the `copyFrom` methods from `Product` so it no longer depends on DTOs.
The conversion now happens in `ProductServiceImpl` using mappers, and the entity exposes clear mutator methods (applyUpdate, applyPatch) for state changes.
This keeps DTO handling in the service layer and makes the domain model cleaner. 😊

Wordbe commented on src/main/kotlin/ecommerce/model/CartItemResponseDTO.kt:
```diff
@@ -5,7 +5,7 @@ import java.time.LocalDateTime
 data class CartItemResponseDTO(
```
What’s the reason for using a `data class` here? 🧐

deniz-oezdemir replied:
Thank you for your question! We used a `data class` here because its primary purpose is to hold data, which is exactly the role of this DTO.

It provides implementations for `equals()`, `hashCode()`, and `toString()`. These are useful for comparisons in tests and for logging during debug sessions. A regular class would require writing this boilerplate manually.

Generally, we use a data class when the primary purpose is to transparently hold data and use a regular class when we care about an object's unique identity, its behavior, or its place in an inheritance structure. Would you agree?

Specifically, would you recommend to use another kind of class or just a regular class here instead? 🤔

Wordbe replied:
I asked because I was wondering if we could just use a regular class instead of a data class. 🙂
It's because `toString()`, `equals()`, `hashCode()`, and `copy()` don’t seem necessary here.

deniz-oezdemir replied:
That's a fair point. I have updated the DTOs to be regular classes as you suggested. This simplifies them where the extra methods are not needed. 👍

Wordbe commented on src/main/kotlin/ecommerce/model/PageResponseDTO.kt:
```diff
@@ -7,9 +7,9 @@ data class PageResponseDTO<T>(
     val numberOfElements: Int,
-    val sort: Any?, // or define your own SortResponse if needed
-    val pageable: Any?, // or define PageableResponse
+    val sort: Any?,
+    val pageable: Any?,
```
Why did you use the `Any` type here? 🙂

deniz-oezdemir replied:
Good catch! We were modeling the response for the tests and used `Any?` as a placeholder. I've now created specific `PageableDTO` and `SortDTO` classes to provide a clear and type-safe contract. Thank you for the suggestion. 🙏

Wordbe commented on src/main/kotlin/ecommerce/model/ProductPatchDTO.kt:
```diff
@@ -7,19 +7,12 @@ import jakarta.validation.constraints.Size
 
 data class ProductPatchDTO(
```
.

deniz-oezdemir replied:
Please refer to my [comment](https://github.com/herotechcourse/spring-ecommerce-order/pull/8#discussion_r2252442544) above. 🙏

Wordbe commented on src/main/kotlin/ecommerce/model/ProductRequestDTO.kt:
```diff
@@ -4,19 +4,21 @@ import ecommerce.util.ValidationMessages.IMAGE_FORMAT
 import jakarta.validation.constraints.Size
 
-data class ProductDTO(
+data class ProductRequestDTO(
```
What about using a regular `class` instead?

deniz-oezdemir replied:
Please refer to my [comment](https://github.com/herotechcourse/spring-ecommerce-order/pull/8#discussion_r2252442544) above. 🙏

Wordbe commented on src/main/kotlin/ecommerce/model/ProductResponseDTO.kt:
```diff
@@ -0,0 +1,31 @@
+import jakarta.validation.constraints.Positive
+import jakarta.validation.constraints.Size
+
+data class ProductResponseDTO(
```
.

deniz-oezdemir replied:
Please refer to my [comment](https://github.com/herotechcourse/spring-ecommerce-order/pull/8#discussion_r2252442544) above. 🙏

Wordbe commented on src/main/kotlin/ecommerce/model/ProductResponseDTO.kt:
```diff
@@ -0,0 +1,31 @@
+import jakarta.validation.constraints.Size
+
+data class ProductResponseDTO(
+    var id: Long? = null,
+    @field:NotBlank(message = NAME_REQUIRED)
+    @field:Size(min = 1, max = 15, message = PRODUCT_NAME_SIZE)
+    @field:Pattern(regexp = "^[a-zA-Z0-9 ()\\[\\]+\\-&/_]*$", message = NAME_PATTERN)
+    var name: String,
+    @field:NotNull(message = PRICE_REQUIRED)
+    @field:Positive(message = PRICE_POSITIVE)
+    var price: Double,
+    @field:NotBlank(message = IMAGE_REQUIRED)
+    @field:Pattern(regexp = "^https?://.*$", message = IMAGE_FORMAT)
+    var imageUrl: String,
```
Why did you use a mutable variable here? Let’s consider the benefits of treating DTOs as immutable.

deniz-oezdemir replied:
You're right, immutability is better for response objects. I've updated the `ProductResponseDTO` to use `val` for its properties. This makes the object's state safer and its purpose as read-only data clearer. Thank you. 👍

Wordbe commented on src/main/kotlin/ecommerce/model/WishItemResponseDTO.kt:
```diff
@@ -5,6 +5,6 @@ import java.time.LocalDateTime
 data class WishItemResponseDTO(
```
.

deniz-oezdemir replied:
Please refer to my [comment](https://github.com/herotechcourse/spring-ecommerce-order/pull/8#discussion_r2252442544) above. 🙏

Wordbe commented on src/main/kotlin/ecommerce/model/WishItemRequestDTO.kt:
```diff
@@ -1,8 +1,5 @@
-import ecommerce.util.ValidationMessages
-import jakarta.validation.constraints.PositiveOrZero
-
 data class WishItemRequestDTO(
```
.

deniz-oezdemir replied:
Please refer to my [comment](https://github.com/herotechcourse/spring-ecommerce-order/pull/8#discussion_r2252442544) above. 🙏

Wordbe commented on src/main/kotlin/ecommerce/services/AdminServiceImpl.kt:
```diff
@@ -1,19 +1,40 @@
 @Service
 @Primary
-class AdminServiceImpl(private val cartItemRepository: CartItemRepository) : AdminService {
+class AdminServiceImpl(
```
Why did you create the `Service` as an interface? Both `Service` and `ServiceImpl` follow this structure, but the implementation doesn't seem to have a distinct name or purpose. 🙂

deniz-oezdemir replied:
That's a great observation. The interface was adding boilerplate without a clear benefit. I have refactored the AdminService to use a single concrete class as you suggested. Thank you. 🙏

Wordbe replied:
If you've decided to remove the interface, how about applying that consistently to other places too, like `ProductService`?

deniz-oezdemir replied:
You're right. For consistency, I should apply this change everywhere. I have now removed the interfaces for the other services as well. 👍

Wordbe commented on src/main/kotlin/ecommerce/services/AuthService.kt:
```diff
@@ -5,19 +5,23 @@ import ecommerce.model.MemberDTO
 
+    @Transactional
     fun createToken(memberDTO: MemberDTO): TokenResponseDTO {
         val accessToken = jwtTokenProvider.createToken(memberDTO.email, memberDTO.role)
```
Why is a transaction needed here? 🙂

deniz-oezdemir replied:
You are right, the transactions here were unnecessary. The database calls are already handled transactionally within the `MemberService`, and `createToken` has no database interaction. I have removed the annotations from `AuthService`. Thank you for pointing that out. 😄

Wordbe commented on src/main/kotlin/ecommerce/services/CartItemServiceImpl.kt:
```diff
@@ -88,10 +97,10 @@ class CartItemServiceImpl(
 
-    override fun deleteAll() {
-        cartItemRepository.deleteAll()
+        existing.quantity = cartItemRequestDTO.quantity
```
How about avoiding setters and getters in `CartItem` and creating methods inside it instead? That way, `CartItem` will have more responsibilities and stronger cohesion. 🙂

deniz-oezdemir replied:
Yes, thanks for pointing that out. We have fixed this already [above](https://github.com/herotechcourse/spring-ecommerce-order/pull/8#discussion_r2251665082). I also removed redundant validation logic from the `CartItemServiceImpl` and now rely on the `CartItem` entity's `updateQuantity` method to enforce its own rules. This improves encapsulation and cohesion. Thank you. 🙏

Wordbe commented on src/main/kotlin/ecommerce/services/ProductServiceImpl.kt:
```diff
@@ -17,50 +17,81 @@ import org.springframework.transaction.annotation.Transactional
+            validateProductNameUniqueness(productDTO.name)
+        }
+
+        existing.copyFrom(productDTO, productDTO.options)
```
Usually, `copyFrom` returns a new object with copied values, like Kotlin data class’s `copy()`. How about using a different method name here? 😀

deniz-oezdemir replied:
That's a great point about method naming. 😊
Confusing copy with update can definitely lead to bugs. The updated implementation uses the name `applyUpdate` for this method to make it clear that it modifies the entity in place. Thank you for ensuring the code is clear and predictable.

Wordbe commented on src/main/kotlin/ecommerce/services/ProductServiceImpl.kt:
```diff
@@ -17,50 +17,81 @@ import org.springframework.transaction.annotation.Transactional
+            validateProductNameUniqueness(productPatchDTO.name!!)
+        }
+
+        existing.copyFrom(productPatchDTO, productPatchDTO.options)
```
Same here.

deniz-oezdemir replied:
I agree completely. The method for applying a patch is now named `applyPatchFromDTO` to make its behavior clear. Thank you for the follow-up.

Wordbe commented on src/test/kotlin/ecommerce/endtoend/AdminE2ETest.kt:
```diff
@@ -12,13 +15,18 @@ import org.junit.jupiter.api.Test
+import kotlin.jvm.java
 
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.DEFINED_PORT)
+@DirtiesContext(classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
```
Since the performance isn’t good, how about trying a different approach? 🚀

deniz-oezdemir replied:
You are right, the performance with `@DirtiesContext` is bad.

In the Admin and WishItem E2E-Tests I've replaced `@DirtiesContext` with a database reset in an `@AfterEach` block that cleans and re-seeds the database between tests via `DatabaseSeeder`. This is much faster.

Thank you for the suggestion. 🙏

Wordbe reviewed (changes requested):
Hello Deniz,

Thank you for learning how to use JPA in a Spring project, writing your code carefully, and submitting it. Great job!

I’ve left review comments, so please check them and apply any possible improvements.

Thank you! 😊

deniz-oezdemir committed:
refactor(api-error): unify error responses using ApiError class
- Create ApiError to standardize error response format
- Update all exception handlers to return ApiError
- Add support for validation errors in the details field
- Add a fallback exception handler
- Change tests to check validation errors in details instead of message

deniz-oezdemir committed:
refactor(endpoints): standardize API endpoints naming
- Renamed POST `/newOption` to `/options` for consistency
- Rename `/api/cart` to `/api/carts`
- Rename `/api/wish` to `/api/wish-items`
- Updated related tests to match new endpoints

deniz-oezdemir committed:
feat(option-validation): enforce productId requirement only for standalone option creation
- Add validation to require productId when creating an option outside of product creation flow
- Update ApiErrorControllerAdvice to handle MissingProductIdException and return 400 Bad Request
- Add E2E tests for missing productId in standalone option creation

deniz-oezdemir committed:
feat(products): validate allowed sort fields for product listing
- Introduce `ALLOWED_SORT_FIELDS` constant to define permitted sort properties
- Add `validateSort` utility in `ProductController` to enforce allowed sort fields

deniz-oezdemir committed:
refactor(CartItem): replace direct quantity assignment with updateQuantity() method
- Add `updateQuantity()`` in CartItem to encapsulate quantity changes
- Update mapping logic to use updateQuantity() instead of direct property set

deniz-oezdemir committed:
refactor(option): add explicit update methods for name and quantity with validation
- Replace direct property setters with `updateName()` and `updateQuantity()` methods in Option entity
- Validate values inside these update methods to improve clarity and maintainability
- Update `mapOptionDTOs()` to use the new update methods instead of setters

deniz-oezdemir committed:
refactor(option): replace magic numbers with named constants
- Introduce MAX_NAME_LENGTH, MIN_QUANTITY, and MAX_QUANTITY constants
- Update validations and annotations to use these constants
- Update test to use MAX_QUANTITY style number

deniz-oezdemir committed:
fix(product): ensure bidirectional association when adding options
- Set `option.product = this` in `addOption` to keep both sides of the
`Product`–`Option` relationship in sync

deniz-oezdemir committed:
refactor(product): remove DTO dependency from Product entity
- Move ProductRequestDTO and ProductPatchDTO conversion logic out of the Product entity
- Add mappers and explicit mutator methods (applyUpdate, applyPatchFromDTO)
- Update ProductServiceImpl to assemble Product and options externally before persisting

deniz-oezdemir committed:
refactor(model): provide specific types for PageResponseDTO
- Add specific `PageableDTO` and `SortDTO` data classes to model the structure of Spring's pagination response

deniz-oezdemir committed:
refactor(model): make ProductResponseDTO properties immutable

deniz-oezdemir committed:
refactor(AdminService): simplify service by removing unnecessary interface

deniz-oezdemir committed:
refactor(AuthService): remove unnecessary @Transactional annotations

deniz-oezdemir committed:
refactor(CartItem): centralize quantity validation in the entity

deniz-oezdemir committed:
refactor(e2e-tests): replace @DirtiesContext with DB reset via `DatabaseSeeder`

deniz-oezdemir commented:
Hello @Wordbe,

Thank you for your thorough and insightful review. I’ve learned a lot from your feedback. 😊 

I have implemented all of the change requests. The tests are more reliable and much faster, the services have been simplified, and the DTOs are more robust. The build is passing, and I believe the PR is in a much better state.

Before you give it a final look, I just wanted to highlight the two discussion points I left unresolved to get your final opinion:
- [On `Cascade.ALL`](https://github.com/herotechcourse/spring-ecommerce-order/pull/8#discussion_r2251725729): In the Product entity, I left a comment explaining my reasoning for using Cascade.ALL, to automatically manage the lifecycle of child entities like options. I just wanted to confirm if my understanding of this best practice is correct for our project.
- [On `data class` vs. `class` for DTOs](https://github.com/herotechcourse/spring-ecommerce-order/pull/8#discussion_r2252442544): I also detailed my thoughts on why data class seems like the right choice for our DTOs. I would appreciate your final opinion on whether you agree with this approach or if you would recommend a regular class in any of these cases.

The pull request is ready for another look. Thank you for your time! 🙏 

Best,
Deniz

deniz-oezdemir committed:
style(): apply automatic code formatting with ktlintFormat

Wordbe reviewed (approved):
Hello Deniz!

Thank you for carefully reading through my review, thinking it over, and applying the feedback thoughtfully.
It wasn’t a small amount of content, so I really appreciate your effort in working through it to the end.

Here are answers to your questions:

1. The `Cascade` option is powerful in that it lets you manage data like objects, but if not used carefully, it can cause unexpected queries. It also tightly couples one object with another. So while there are trade-offs, it can be very useful when used wisely.
2. Regarding the `data class`, I mentioned it in the comments. I generally avoid it unless necessary, as it generates methods I may not need. I suggested using it only when those methods are truly useful.

You’ve reflected most of the feedback very well, so I’ll approve and merge this PR now.
Please carry over any unfinished improvements into the next step.

Great job, and thank you! 🙌
